File: app/src/main/java/br/com/alura/dao/AlunoDAO.java
package br.com.alura.dao;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.NonNull;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import br.com.alura.modelo.Aluno;

public class AlunoDAO extends SQLiteOpenHelper {

    private static final String DB_NAME = "Agenda";
    private static final int DB_VERSION = 2;

    public AlunoDAO (Context context) {
        super(context, DB_NAME, null, DB_VERSION);
    }

    @Override
    public void onCreate (SQLiteDatabase db) {
        String sql = "CREATE TABLE Alunos (id INTEGER PRIMARY KEY, nome TEXT NOT NULL, endereco TEXT, telefone TEXT, site TEXT, nota REAL);";
        db.execSQL(sql);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        String sql = "DROP TABLE IF EXISTS Alunos";
        db.execSQL(sql);
        onCreate(db);
    }

    public void insere(@NotNull Aluno aluno)  {
        SQLiteDatabase db = getWritableDatabase();

        ContentValues dados = pegaDadosDoAluno(aluno);

        db.insert("Alunos", null, dados );
    }

    public List<Aluno> buscaAlunos()  {
        String sql = "SELECT * FROM Alunos;";
        SQLiteDatabase db = getReadableDatabase();
        Cursor c = db.rawQuery(sql, null);

        List<Aluno> alunos = new ArrayList<Aluno>();
        while (c.moveToNext()) {
            Long id = c.getLong(c.getColumnIndex("id"));
            String nome = c.getString(c.getColumnIndex("nome"));
            String endereco = c.getString(c.getColumnIndex("endereco"));
            String telefone = c.getString(c.getColumnIndex("telefone"));
            String site = c.getString(c.getColumnIndex("site"));
            Double nota = c.getDouble(c.getColumnIndex("nota"));
            Aluno aluno = new Aluno(id, nome, endereco, telefone, site, nota);
            alunos.add(aluno);
        }
        c.close();

        return alunos;

    }

    public void remover(@NotNull Aluno aluno) {
        SQLiteDatabase db = getWritableDatabase();

        String [] params = {String.valueOf(aluno.getId())};
        db.delete("Alunos", "id = ?", params);
    }

    public void altera(@NotNull Aluno aluno) {
        SQLiteDatabase db = getWritableDatabase();

        ContentValues dados = pegaDadosDoAluno(aluno);

        String[] params ={aluno.getId().toString()};
        db.update("Alunos", dados, "id = ?", params);
    }

    @NonNull
    private ContentValues pegaDadosDoAluno(Aluno aluno) {
        ContentValues dados = new ContentValues();
        dados.put("nome", aluno.getNome()); dados.put("endereco", aluno.getEndereco());
        dados.put("telefone", aluno.getTelefone());
        dados.put("site", aluno.getSite());
        dados.put("nota", aluno.getNota());
        return dados;
    }
}

File: app/src/main/java/br/com/alura/dao/AlunoDAO.kt
package br.com.alura.dao

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import br.com.alura.modelo.Aluno
import java.util.ArrayList

class AlunoDAO(context: Context): SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {

    companion object {
        private val DB_NAME = "Agenda"
        private val DB_VERSION = 3
    }

    override fun onCreate(db: SQLiteDatabase?) {
        val sql = "CREATE TABLE Alunos (" +
                    "id INTEGER PRIMARY KEY, " +
                    "nome TEXT NOT NULL, " +
                    "endereco TEXT, " +
                    "telefone TEXT, " +
                    "site TEXT, " +
                    "nota REAL, " +
                    "foto TEXT);"
        db?.execSQL(sql)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        val sql: String =
            when(oldVersion){
                2,3 -> "ALTER TABLE Alunos ADD COLUMN foto TEXT"
                else -> ""
            }

        db?.execSQL(sql)
    }

    fun insere(aluno: Aluno) {
        val db = writableDatabase

        val dados = pegaDadosDoAluno(aluno)

        db.insert("Alunos", null, dados)
    }

    fun buscaAlunos(): List<Aluno> {
        val sql = "SELECT * FROM Alunos;"
        val db = readableDatabase
        val c = db.rawQuery(sql, null)

        val alunos = ArrayList<Aluno>()
        while (c.moveToNext()) {
            val id = c.getLong(c.getColumnIndex("id"))
            val nome = c.getString(c.getColumnIndex("nome"))
            val endereco = c.getString(c.getColumnIndex("endereco"))
            val telefone = c.getString(c.getColumnIndex("telefone"))
            val site = c.getString(c.getColumnIndex("site"))
            val nota = c.getDouble(c.getColumnIndex("nota"))
            val foto = c.getString(c.getColumnIndex("foto"))
            val aluno = Aluno(id, nome, endereco, telefone, site, nota, foto)
            alunos.add(aluno)
        }
        c.close()

        return alunos

    }

    fun remover(aluno: Aluno) {
        val db = writableDatabase

        val params = arrayOf(aluno.id.toString())
        db.delete("Alunos", "id = ?", params)
    }

    fun altera(id: Long, aluno: Aluno) {
        val db = writableDatabase

        val dados = pegaDadosDoAluno(aluno)

        val params = arrayOf(id.toString())
        db.update("Alunos", dados, "id = ?", params)
    }

    private fun pegaDadosDoAluno(aluno: Aluno): ContentValues {
        val dados = ContentValues()
        dados.put("nome", aluno.nome)
        dados.put("endereco", aluno.endereco)
        dados.put("telefone", aluno.telefone)
        dados.put("site", aluno.site)
        dados.put("nota", aluno.nota)
        dados.put("foto", aluno.foto)
        return dados
    }
}